Skip tests if socket buffers are inadequate

wferi opened the pull request:
What do you think about this one instead of #97?
It depends on `ERANGE` having a fixed semantics from `knet_handle_new()`.

fabbione commented:
Hey Feri,

I start to like the idea to use a dedicated errno for this check, but we should probably avoid ENORANGE. I know it makes sense because it is indeed a value problem, but it might be masking other calls in the code path that can return ENORANGE for a good failure that we want to catch during testing.

What about something a bit absurd like ENAMETOOLONG?

The next thing is the coding style. I know many love this syntax that you are using. It´s not used anywhere in knet :-) I would appreciate if you could align it.

Thanks
Fabio

wferi commented:
I've got nothing against `ENAMETOOLONG`, that even makes some sense. But cleanly separating OS and knet error codes would be more future-proof. And we've already got a handle around, which could store arbitrarily detailed error information. Check out this [StackOverflow answer](https://stackoverflow.com/a/14551626/1142595) for a worked-out example (or the others for different ideas).

wferi commented:
Concerning my style, I can't see anything outstanding in this patch. Are you annoyed by the occasional spaces after the function names I smuggled in with earlier commits? I'll try to suppress them completely.

fabbione commented:
For the coding style:

 +		int exit_status = errno == ERANGE ? SKIP : FAIL; 

I am not a big fan of the ? : : format. I prefer to see explicit if () {} else {}.

Just come to my mind that this change affects all tests and it is specific to the test suite, it might make sense to have a wrapper around opening a knet_handle.

fabbione commented:
Nice idea around error handling. Let´s get in for 2.0 (it will require API/ABI changes across the board).

fabbione commented:
https://trello.com/c/hbzrZ0dI/799-review-knet-error-report-to-separate-os-errors-from-internal-knet-error

wferi committed:
Report ENAMETOOLONG if socket buffers are too small

Signed-off-by: Ferenc Wágner <wferi@debian.org>

wferi committed:
tests: introduce knet_handle_start helper

Signed-off-by: Ferenc Wágner <wferi@debian.org>

wferi committed:
tests: replace open coded error handling with knet_handle_start() calls

Done by:
$ sed -i -e '1h;2,$H;$!d;g' -e 's/knet_h = knet_handle_new(1, logfds\[1\], \(KNET_LOG_[A-Z]*\));\n\n\tif (!knet_h) {\n\t\tprintf("knet_handle_new failed: %s\\n", strerror(errno));\n\t\tflush_logs(logfds\[0\], stdout);\n\t\tclose_logpipes(logfds);\n\t\texit(FAIL);\n\t}/knet_h = knet_handle_start(logfds, \1);/g' libknet/tests/*.c

Signed-off-by: Ferenc Wágner <wferi@debian.org>

wferi committed:
tests: skip if socket buffers are too small

Signed-off-by: Ferenc Wágner <wferi@debian.org>

fabbione commented:
Hi Feri, nice work. I just noticed a couple of bits that we should handle better. Sorry I didn´t see them before.

        if (setsockopt(sock, SOL_SOCKET, option, &target, sizeof target) == 0 &&
            getsockopt(sock, SOL_SOCKET, option, &new_value, &value_len) == 0) {

^^^^ here we should capture the *etsockopt errno and report it in the logs. I think it´s important we get to see why those calls are failing otherwise debugging might be complicated.

                if (value_len == sizeof new_value && target <= new_value) {
                        return 0;
                }
                if (!force) {
                        log_debug (knet_h, KNET_SUB_TRANSPORT,
                                   "Failed to set socket buffer via option %d to value %d: capped at %d",
                                   option, target, new_value);
                        errno = ENAMETOOLONG;
                        return -1;
                }
        }

Now that we are effectively remapping an error code, we should document it in libknet.h too so that users know what it means and how to handle it.
don´t worry about updating the man pages. If we get to merge this before 1.0, I´ll do a batch update before release.

wferi committed:
More fine-grained error handling

Signed-off-by: Ferenc Wágner <wferi@debian.org>

wferi committed:
Document knet-specific errno values

Signed-off-by: Ferenc Wágner <wferi@debian.org>

wferi commented:
Something like this? I'm not sure about the severities, though, it's a mix of `log_err` and `log_debug` now.
Also, shall we print the explanation before `exit(SKIP)` in the generic case as well?

fabbione commented:
Yeah it looks pretty good. Anything that will stop operation and return an error should be log_err(. See _configure_transport_socket as an example.

fabbione commented:
Also, while at it, can you please remove the space between function_name and ( ?
I prefer function_name(.. vs "function_name (". I spotted in a few places around ;)

Thanks
Fabio

wferi committed:
Anything that stops operation and returns an error should be log_err

Signed-off-by: Ferenc Wágner <wferi@debian.org>

wferi committed:
Remove spaces after function names

Signed-off-by: Ferenc Wágner <wferi@debian.org>
